client/lib/plugins/sanitize-section-content.js

```diff
@@ -0,0 +1,246 @@
+ * @param {String} attrName name of attribute under inspection
+ * @returns {Boolean} whether the attribute is allowed
+ */
+const isAllowedAttr = ( tagName, attrName ) => {
+	switch ( tagName ) {
+		case 'a':
+			return 'href' === attrName;
+
+		case 'iframe':
+			switch ( attrName ) {
+				case 'class':
+				case 'type':
+				case 'height':
+				case 'width':
+				case 'src':
+					return true;
+				default:
+					return false;
+			}
+
+		case 'img':
+			switch ( attrName ) {
+				case 'alt':
+				case 'src':
+					return true;
+				default:
+					return false;
+			}
+
+		default:
+			return false;
+	}
+};
+
+const isValidYoutubeEmbed = node => {
+	if ( node.nodeName.toLowerCase() !== 'iframe' ) {
+		return false;
+	}
+
+	if ( node.getAttribute( 'class' ) !== 'youtube-player' ) {
+		return false;
+	}
+
+	if ( node.getAttribute( 'type' ) !== 'text/html' ) {
+		return false;
+	}
+
+	const link = document.createElement( 'a' );
+	link.href = node.getAttribute( 'src' );
+
+	return (
+		validUrl.isWebUri( node.getAttribute( 'src' ) ) &&
+		( link.hostname === 'youtube.com' || link.hostname === 'www.youtube.com' )
+	);
+};
+
+const replacementFor = node => {
+	const tagName = node.nodeName.toLocaleLowerCase();
+
+	switch ( tagName ) {
+		case 'h1':
+		case 'h2':
+			return 'h3';
+
+		default:
+			return null;
+	}
+};
+
+/**
+ * Sanitizes input HTML for security and styling
+ *
+ * @param {String} content unverified HTML
+ * @returns {string} sanitized HTML
+ */
+export const sanitizeSectionContent = content => {
+	const parser = new DOMParser();
+	const doc = parser.parseFromString( content, 'text/html' );
+
+	// this will let us visit every single DOM node programmatically
+	const walker = doc.createTreeWalker( doc.body );
+
+	/**
+	 * we don't want to remove nodes while walking the tree
+	 * or we'll invite data-race bugs. instead, we'll track
+	 * which ones we want to remove then drop them at the end
+	 *
+	 * @type {Array<Node>} List of nodes to remove
+	 */
+	const removeList = [];
+
+	/**
+	 * track any tags we want to replace, we'll need to
+	 * transfer the children too when we do the swap
+	 *
+	 * @type {Array<Node, Node>} List of pairs of nodes and their replacements
+	 */
+	const replacements = [];
+
+	// walk over every DOM node
+	while ( walker.nextNode() ) {
+		const node = walker.currentNode;
+		const tagName = node.nodeName.toLowerCase();
+
+		if ( ! isAllowedTag( tagName ) && ! isValidYoutubeEmbed( node ) ) {
+			removeList.push( node );
+			continue;
+		}
+
+		const replacement = replacementFor( node );
+		if ( replacement ) {
+			replacements.push( [ node, document.createElement( replacement ) ] );
+		}
+
+		// strip out anything not explicitly allowed
+		// in the attributes. we want to eliminate
+		// potential cross-site scripting attacks _and_
+		// prevent custom styles from interfering with
+		// our page's own rendering
+
+		const attrRemoveList = [];
+
+		// these aren't Arrays, so iteration is odd
```

Comment:
This comment can be removed -- the iteration is no longer odd 😄

Reply:
I'd like to leave it in here still because it _seems_ like we could just use `Array.prototype.forEach` here but we _can't_. I'd rather leave in the explanation of why it's odd than leave the next developer wondering why it breaks when they come in and "fix" it, but maybe I should clarify in the comment the _why_…